torch/fx/passes/infra/partitioner.py

```diff
@@ -0,0 +1,227 @@
+        # -1 if a depends on b (,or equivalently b is an upstream depedency of a)
+        # 0 if a and b doesn't have dependency between each other
+
+        # TODO: build a cache here to speedup the query
+
+        for node_a in partition_a.nodes:
+            for node_b in partition_b.nodes:
+                dependency = self.__node_depends_on(node_a, node_b)
+                if dependency != 0:
+                    return dependency
+        return 0
+
+    def __get_supported_nodes(self) -> NodeList:
+        logging.debug("Collecting supported nodes...")
+        supported_nodes = []
+        for node in self.graph_module.graph.nodes:
+            if self.operator_support.is_node_supported(self.graph_module.named_modules(), node):
+                supported_nodes.append(node)
+        return supported_nodes
+
+    def propose_partitions(self) -> List[Partition]:
+        candidates: NodeList = self.__get_supported_nodes()
+
+        # assumptions: nodes in candidate list is sorted in topological order
+        assignment: Dict[Node, int] = {}   # maping from node to partition_id
+        partitions_by_id: Dict[int, Partition] = {}  # mapping from partition_id to partition
+        new_partition_id = itertools.count()
+
+        def assign(node: Node, id: Optional[int] = None):
+            # If id is None, remove the node from original assigment
+
+            # node has been assigned before, clean up and re-assign
+            if node in assignment:
+                original_id = assignment[node]
+                del assignment[node]
+                partitions_by_id[original_id].remove_node(node)
+                if partitions_by_id[original_id].size() == 0:
+                    del partitions_by_id[original_id]
+
+            if id is not None:
+                assignment[node] = id
+                if id not in partitions_by_id:
+                    partitions_by_id[id] = Partition(id=id, nodes=[node])
+                else:
+                    partitions_by_id[id].add_node(node)
+
+        logging.debug("Proposing partitions...")
+
+        # visit candidates in reversed topological order
+        for node in reversed(candidates):
+            # use Dict as an ordered set to ensure deterministic partitioning result, don't care value
+            user_partitions: Dict[Partition, None] = {}
+            for user_node in node.users:
+                if user_node in assignment:
+                    id = assignment[user_node]
+                    user_partitions[partitions_by_id[id]] = None
+                else:
+                    user_partitions[Partition(nodes=[user_node])] = None
+
+            # Filter out all the partitions that has dependency on other users
+            # TODO: find a better way to do this, rather than pair-wise comparision
+            user_partitions_list = list(user_partitions.keys())
+            for i in range(len(user_partitions_list)):
```

Comment:
Consider following example:
node has 3 user partitions.
p1 depends on p0, so node cannot be fused into p1, otherwise it will introduce a dependency circle. 
node can be fuse into either p0 or p2. so after this loop, assigned_candidate_partition_ids would only contain [0, 2]. 
In this case, [node, p0, and p2] can all be fused into a single partition. This is handled at line 164. 


![image](https://user-images.githubusercontent.com/9906745/175644486-fef6ec06-9ee0-4848-8e5f-16d0ee52defa.png)